README.md
# MYZS (My Z Shell)

This is a configable rc file (shell initial file) mainly support `zshrc`, with possible to support bash.

- [Development](./docs/DEV.md)
- [Utilities](./docs/README_UTILS.md)
- [Plugins](./docs/README_PLUGINS.md)
- [Changelog](./docs/CHANGELOG.md)
- [Line of Code](./docs/LOC.md)

## Usage

This project wroted all with shell script. 
No installation require and support MacOS and Linux and Window (with `Windows Subsystem for Linux` installed).
You need to copy the application place to correct location and every work fine. Always check configuration before use it because configuration might not suit for you ([config](#configuration)).

### Automatic

1. Open terminal / iTerm 2
2. Run command `source <(curl -sNL https://github.com/kamontat/myzs/raw/master/scripts/global.sh)`
3. Answer all question and all set

### Manually

1. Fork the project
2. Clone project via `git clone $HOME/.myzs` or download by Github UI
3. Link `.zshrc` from myzs to root $HOME via `ln -s "$HOME/.myzs/.zshrc" "$HOME/.zshrc"`.

### Command

utils script for manage this project.

1. `./scripts/command.sh install` - This will install command to current .zshrc files
2. `./scripts/command.sh uninstall` - This will uninstall this project add delete all file

## Configuration

All configuration already listed in `.zshrc` file.

### Tips

I attach a lot of features including debugging so by default I might enable some settings / modules depend on my personal needs. This is a list of features you might set differently to improve load time. Your point of view is on `$MYZS_LOADING_SETTINGS` and `$MYZS_LOADING_MODULES`

1. progressbar - by default I will always enabled progressbar to improve UX (improving ~7.3%)
   1. aggregate data - plugins and modules tend to have many of it, we can aggregate result and report once with setting below (improving ~30.4%)

```bash
"$" disabled pg # MYZS_LOADING_SETTINGS
"$" enabled myzs/plugin/aggregation
"$" enabled myzs/module/aggregation
```

2. metrics - currently metric apis subject to changes a lot so now it not improve anything yet

```bash
"$" disabled metric # MYZS_LOADING_SETTINGS
```

3. logging - on normal situation only error log should enough to make script works and idenify problem if any (improving ~7.8%)

```bash
"$" array logger/level "error" # MYZS_LOADING_SETTINGS
```

## Dependencies

1. [zplug](https://github.com/zplug/zplug): for plugin manangement
2. [gitgo](https://github.com/kamontat/gitgo): for git commit management


CHANGELOG.md
# Release notes

## Version 5.6.3 (09 Jun 2021)

- reduce directory check on plugins
- add new error log when validate module failed
- remove variable cleanup because it might use via `myzs-*` commands
- add new apis `_myzs:internal:call-or` for calling if internal command exist or fallback command
- reduce more debug log on low level checker
- auto create directory before anything
- change zplug plugin to zplugin and myzs-plugin to plugin to avoid confuse
- now we can append array in database data via `_myzs:internal:db:append:array`
- implement plugins system to depend on database apis (similar with setting)
- due to migration plugins system, plugin metrics will disabled for a while
- improve how plugin method works to simpify custom steps
- update LOC.md

| Type                               | Tasks | Load time |
| ---------------------------------- | ----- | --------- |
| Default config                     | 28    | 5s:122ms  |
| Enable all log level               | 28    | 5s:607ms  |
| Enable performance log             | 28    | 6s:702ms  |
| Enabled all log and performance    | 28    | 7s:166ms  |
| Enabled all available myzs-plugins | 43    | 7s:368ms  |
| Enabled all features and plugins   | 43    | 10s:474ms |

## Version 5.5.2 (07 Jun 2021)

- add `myzs-debug` with same parameters with myzs-measure, it will print all executed commands
- improve `myzs-measure` to expose error code to exit as well
- improve progressbar on success, skip, and failure case
- fix manpage error on macos
- remove duplicate setup with autocd module
- reduce output length on several place (e.g. progressbar, debug log, or error message)
- remove some useless debug log to improve load time
- remove initial `init file` when load plugin, since we didn't use it anywhere

| Type                                 | Load time |
| ------------------------------------ | --------- |
| All features and plugins             | 13s:148ms |
| Exclude performance                  | 9s:177ms  |
| enabled only error from above config | 8s:667ms  |
| Enabled 3 plugins from above config  | 4s:723ms  |

## Version 5.5.1 (05 Jun 2021)

- reduce 5 builtin modules and create separate plugins instead
- add `app/autocd` module for autoload myzs-setup file if exist when cd to directory
- remove all low-level log, this reduce loadtime up to 35%
- update default values in **.zshrc** file
- not enabled **app/vscode** and **app/git** modules by default
- document database apis and settings apis

## Version 5.5.0 (05 Jun 2021)

- add new database apis in utils for create, update data using in myzs
- move env support as builtin module (you can disable via remove from loading module)
- add new command `myzs-loadtest <n> <cmd...>` for measure and calculate average timing (ms)
- if you didn't provide myzs type, by default it will loading every specify modules
- migrated new settings apis to database apis and deprecate old setting version (will remove soon)
- change golang default path to **$HOME/go**
- avoid calling unknown method when setup setting command type
- fix myzs-info report wrong loading duration when running myzs-measure
- fix revolver throw wrong exception when pg/style is invalid
- upgrade gitgo from v4 to v5 to support new features
- create poc module group (not working)

## Version 5.4.0 (10 May 2021)

- remove MYZS_DEBUG since it useless on debugging (too many log)
- do cleanup on each module loaded instead
- `_myzs:internal:module:cleanup` is accept optional module_key
- notice that `__MYZS__CURRENT_MODULE_*` will exist until end of loading only (not carry over)

## Version 5.3.5 (10 May 2021)

- wrong License type on `myzs-info` command (it should be **AGPL-3.0**, not **MIT**)

## Version 5.3.4 (10 May 2021)

- add logging when checking os and shell
- upgrade app/gcloud.sh to support latest version
- add new plugin `myzs-plugins/python` for conda support
- now plugin default branch will be **main** instead of **master**

## Version 5.3.3 (21 Apr 2021)

- add error progress when zplug load failed
- add error progress when configured zplug failed
- add error when loading module with error and `pb/performance` is enabled
- change custom zplug plugins list name from `myzs:zplug:plugin-list` to `myzs:zplug:custom:plugin-list`
- fix not color final message from progressbar
- fix `source "$file"` return zero even error occurred

## Version 5.3.2 (21 Apr 2021)

- fix `/tmp/myzs` didn't create for data and log path
- log file will create even never write any log to file

## Version 5.3.1 (16 Apr 2021)

- fix deploy-plugins scripts

## Version 5.3.0 (16 Apr 2021)

- [BREAK] module will not reload if last load is passed
- move `$MYZS_SETTINGS_AUTOLOAD_SETUP_LOCAL`, `$MYZS_SETTINGS_SETUP_FILES` and `$MYZS_LOG_LEVEL` to new settings format
- add new settings `myzs/zplug` for enabled / disabled zplug completely
- change settings `path/auto-open` to `automatic/open-path`
- change default metrics to disabled
- add **alias/git.sh** by default
- remove **src/alias/fuck.sh** use `myzs-plugins/thefuck#master` plugins instead
- remove **src/app/generator.sh** and **src/app/github.sh** use `kamontat/mplugin-kamontat#master` plugins instead
- remove **src/app/fuck.sh**
- remove **src/app/fzf.sh**
- remove **src/app/trivis.sh**
- support settings value as array
- comment a lot of unused zplug plugins
- remove all `_myzs:internal:log:set-level` function
- add new array support apis in settings.sh module
- update helpers apis list
- remove project myzs-setup since git is loaded by default
- remove `$MYZS_TYPE` on plugin template [plugin]
- fix deploy script throw error when beta/alpha version is preset

## Version 5.3.0-beta.2 (15 Apr 2021)

- re-release 5.3.0-beta.1 due to push failed

## Version 5.3.0-beta.1 (15 Apr 2021)

- move settings to single array `${MYZS_LOADING_SETTINGS}`
- add new modules **helper/setting.sh** to manage about all settings
- make all debug log from checker to be lower case

init.sh
# shellcheck disable=SC2148

export _MYZS_ROOT="${MYZS_ROOT:-"$HOME/.myzs"}"
export __MYZS__SRC="${_MYZS_ROOT}/src"
export __MYZS__PLG="${_MYZS_ROOT}/plugins"
export __MYZS__COM="${_MYZS_ROOT}/resources/completion"

export __MYZS__LIB="${__MYZS__SRC}/lib"
export __MYZS__UTL="${__MYZS__SRC}/utils"

export ZPLUG_HOME="${MYZS_ZPLUG:-${_MYZS_ROOT}/zplug}"

export __MYZS__OWNER="Kamontat Chantrachirathumrong"
export __MYZS__VERSION="5.7.2"
export __MYZS__SINCE="21 Apr 2018"
export __MYZS__LAST_UPDATED="12 Jun 2021"
export __MYZS__LICENSE="AGPL-3.0 License"

export __MYZS__MODULES=()

# ########################## #
# Start loading dependencies #
# ########################## #

# shellcheck disable=SC1091
source "${__MYZS__UTL}/index.sh"

_myzs:internal:setting:initial
myzs:module:new "$0"

# load progress bar
if _myzs:internal:checker:fully-type; then
  _myzs:internal:module:load "builtin#lib/progress.sh"
else
  _myzs:internal:module:load "builtin#lib/dump-progress.sh"
fi

myzs:pg:start

# initial system config
myzs:pg:step "Command" "Setup system settings"
_myzs:internal:module:load "builtin#settings/system.sh" || myzs:pg:mark-fail "Commandline" "Cannot load system variable"

# initial path config
myzs:pg:step "Command" "Setup path settings"
_myzs:internal:module:load "builtin#settings/path.sh" || myzs:pg:mark-fail "Commandline" "Cannot load path variable"

# initial zsh config
if _myzs:internal:checker:fully-type && _myzs:internal:checker:shell:zsh; then
  myzs:pg:step "Command" "Setup commandline settings"
  _myzs:internal:module:load "builtin#settings/zsh.sh" || myzs:pg:mark-fail "Commandline" "Cannot load zsh settings"
fi

# load myzs plugins
myzs:pg:step "Plugin" "Initial ${#MYZS_LOADING_PLUGINS[@]} plugins"
for plugin in "${MYZS_LOADING_PLUGINS[@]}"; do
  if _myzs:internal:setting:is-disabled "plugin/aggregation"; then
    myzs:pg:step "Plugin" "Loaded plugin $plugin"
    if ! _myzs:internal:plugin:initial "$plugin"; then
      myzs:pg:mark-fail "Plugin" "Cannot load $plugin plugin"
      continue
    fi
  else
    _myzs:internal:plugin:initial "$plugin"
  fi
done

# load myzs modules
myzs:pg:step "Module" "Initial ${#MYZS_LOADING_MODULES[@]} modules"
for module in "${MYZS_LOADING_MODULES[@]}"; do
  if _myzs:internal:setting:is-disabled "module/aggregation"; then
    myzs:pg:step "Module" "Loaded module $module"
    if ! _myzs:internal:module:load "$module"; then
      myzs:pg:mark-fail "Module" "Cannot load $module module"
      continue
    fi
  else
    _myzs:internal:module:load "$module"
  fi
done

if _myzs:internal:setting:is-enabled "group"; then
  myzs:pg:step "Group" "Initial group data"
  _myzs:internal:group:initial "${MYZS_LOADING_GROUPS[@]}" || myzs:pg:mark-fail "Group" "Cannot load $? group"
fi

# initial zplug config
if _myzs:internal:checker:fully-type &&
  _myzs:internal:checker:shell:zsh &&
  _myzs:internal:setting:is-enabled "zplug"; then
  myzs:pg:step "ZPlugin" "Initial zplug configuration"
  _myzs:internal:module:load "zplug#init.zsh" || myzs:pg:mark-fail "ZPlugin" "Cannot load zplug initial script"

  myzs:pg:step "ZPlugin" "Initial zplug plugin list"
  _myzs:internal:module:load "builtin#settings/zplugins.sh" || myzs:pg:mark-fail "ZPlugin" "Cannot load zplug plugins list"

  myzs:zplug:initial-plugins # initial plugins

  # Install plugins if there are plugins that have not been installed
  if ! zplug check --verbose; then
    printf "Install? [y/N]: "
    if read -rq; then
      echo
      zplug install
    fi
  fi

  myzs:pg:step "ZPlugin" "Loaded zplug plugin list"
  # load new plugins to system
  if ! zplug load --verbose >>"$MYZS_ZPLUG_LOGPATH"; then
    myzs:pg:mark-fail "ZPlugin" "Cannot load zplug plugin"
  fi

  myzs:pg:step "ZPlugin" "Setup zplug plugin config"
  if ! myzs:zplug:setup-plugins; then
    myzs:pg:mark-fail "ZPlugin" "Cannot configure zplug plugin"
  fi
fi

myzs:pg:stop
